diff -uNr nginx-1.2.2/auto/cc/gcc nginx-1.2.2.new/auto/cc/gcc
--- nginx-1.2.2/auto/cc/gcc	2012-01-18 23:07:43.000000000 +0800
+++ nginx-1.2.2.new/auto/cc/gcc	2013-11-01 16:19:19.357494691 +0800
@@ -170,7 +170,7 @@
 
 
 # stop on warning
-CFLAGS="$CFLAGS -Werror"
+#CFLAGS="$CFLAGS -Werror"
 
 # debug
 CFLAGS="$CFLAGS -g"
diff -uNr nginx-1.2.2/auto/cc/icc nginx-1.2.2.new/auto/cc/icc
--- nginx-1.2.2/auto/cc/icc	2012-01-18 23:07:43.000000000 +0800
+++ nginx-1.2.2.new/auto/cc/icc	2013-11-01 16:19:35.805495168 +0800
@@ -115,7 +115,7 @@
 esac
 
 # stop on warning
-CFLAGS="$CFLAGS -Werror"
+#CFLAGS="$CFLAGS -Werror"
 
 # debug
 CFLAGS="$CFLAGS -g"
